### ISSUE TITLE
AttributeError: module 'ssl' has no attribute 'wrap_socket'

### ISSUE THREAD

**Issue by MarkStoutjesdijk** [supersaiyanmode/PyWebOSTV#108]
Hi,

Using secure connection, I am getting an error:

AttributeError: module 'ssl' has no attribute 'wrap_socket'

I suspect it is due to a change in Python -- see https://github.com/eventlet/eventlet/issues/795. It would be great if you could take a look at this!

**Comment by supersaiyanmode** [comment]
Thanks! I didn't know about this one. This is indeed because ssl library changes in latest Python.

I was looking [at the connect(..) here](https://github.com/Lawouach/WebSocket-for-Python/blob/961c07ce16ce4eedc34ca1fdacd29442870feccc/ws4py/client/__init__.py#L214) -- unfortunately, it looks like it hasn't been written in a manner that allows for SSL context to overridden by child classes, without copying the entire function.

That being said, we can perhaps try having `ssl.wrap_socket` point to something else ( **HACKY + UNTESTED**):

```
# Before any library imports that depends on ssl..
ssl.wrap_socket = ssl.SSLContext().wrap_socket
```

.. can you please confirm if this works for you? If so, I'll try incorporating this in this library..

**Comment by MarkStoutjesdijk** [comment]
<div></div><div>Thanks! I didn't know about this one. This is indeed because ssl library changes in latest Python.
<br/>
<br/>I was looking [at the connect(..) here](https://github.com/Lawouach/WebSocket-for-Python/blob/961c07ce16ce4eedc34ca1fdacd29442870feccc/ws4py/client/__init__.py#L214) -- unfortunately, it looks like it hasn't been written in a manner that allows for SSL context to overridden by child classes, without copying the entire function.
<br/>
<br/>That being said, we can perhaps try having `ssl.wrap_socket` point to something else ( **HACKY + UNTESTED**):
<br/>
<br/>```
<br/># Before any library imports that depends on ssl..
<br/>ssl.wrap_socket = ssl.SSLContext().wrap_socket
<br/>```
<br/>
<br/>.. can you please confirm if this works for you? If so, I'll try incorporating this in this library..
<br/>
<br/>-- 
<br/>Reply to this email directly or view it on GitHub:
<br/>https://github.com/supersaiyanmode/PyWebOSTV/issues/108#issuecomment-2095947350
<br/>You are receiving this because you authored the thread.
<br/>
<br/>Message ID: ***@***.***&gt;</div>

**Comment by TJennerjahn** [comment]
Came here because of that exact issue. Can confirm that this:
```python
ssl.wrap_socket = ssl.SSLContext().wrap_socket
```

does fix it. Not great as a permanent solution, but I'll take. Thanks :)

**Comment by umetzu** [comment]
Yes, his solution works,  add the protocol to  skip the deprecation warning.

ssl.wrap_socket = ssl.SSLContext(protocol=ssl.PROTOCOL_TLSv1_2).wrap_socket

**Comment by apppie123** [comment]
Whas this already implemented? I just installed 8.10 but I am getting this error.

Exception has occurred: AttributeError
module 'ssl' has no attribute 'wrap_socket'
  File "src/test.py", line 64, in <module>
    client.connect()
AttributeError: module 'ssl' has no attribute 'wrap_socket'

**Comment by hawaii67** [comment]
This is due to Python 3.12 having removed the ssl.wrap_socket() function (see
https://docs.python.org/3.12/whatsnew/3.12.html#ssl).

**Comment by DoughMucker** [comment]
> Yes, his solution works, add the protocol to skip the deprecation warning.
> 
> ssl.wrap_socket = ssl.SSLContext(protocol=ssl.PROTOCOL_TLSv1_2).wrap_socket


Hi, Could you explain which file/line this should be inserted into? I tried a few places, but it doesn't seem to work for me.
Thanks.

NVM, figured it out. Had to insert it at line 214 of ws4py/client/__init__.py
